crates/bifrost/src/read_stream.rs

```diff
@@ -11,12 +11,155 @@
 use std::sync::Arc;
 
 use crate::bifrost::BifrostInner;
-use crate::{LogId, Lsn};
+use crate::{Error, LogId, LogRecord, Lsn};
 
-pub struct LogReadStream {}
+pub struct LogReadStream {
+    inner: Arc<BifrostInner>,
+    log_id: LogId,
+    read_pointer: Lsn,
+}
 
 impl LogReadStream {
-    pub(crate) fn new(_inner: Arc<BifrostInner>, _log_id: LogId, _from: Lsn) -> Self {
-        Self {}
+    pub(crate) fn new(inner: Arc<BifrostInner>, log_id: LogId, after: Lsn) -> Self {
+        Self {
+            inner,
+            log_id,
+            read_pointer: after,
+        }
+    }
+
+    fn seek_to(&mut self, record: &LogRecord) {
+        let read_pointer = match &record.record {
+            // The log record is a gap in the log. we deliver the gap then fast-forward to the
+            // until pointer so that future reads would skip that gap.
+            crate::Record::TrimGap(trim_gap) => trim_gap.until,
+            crate::Record::Data(_) => record.offset,
+            crate::Record::Seal(_) => record.offset,
+        };
+        self.read_pointer = read_pointer;
+    }
+
+    /// Read the next record from the log after the current read pointer. The future will resolve
+    /// after the record is available to read, this will async-block indefinitely if no records are
+    /// ever written to the log beyond the read pointer.
+    ///
+    /// This future is "Cancellation" safe.
+    pub async fn read_next(&mut self) -> Result<LogRecord, Error> {
+        let record = self
+            .inner
+            .read_next_single(self.log_id, self.read_pointer)
+            .await?;
```

Comment:
nit: Depending on how `read_next` is used (e.g. in some select statement), then this might create a couple of `OffsetWatchers` that are only cleaned up once the commit offset goes beyond the `self.read_pointer`. Probably not a problem.

Reply:
That's true in memory loglet which (OffsetWatcher is a memory-loglet thing.